src/main/scala/firrtl/passes/CheckCombLoops.scala

```diff
@@ -0,0 +1,146 @@
+// See LICENSE for license details.
+
+package firrtl.passes
+
+import scala.collection.mutable
+import scala.collection.immutable.HashSet
+import scala.collection.immutable.HashMap
+import annotation.tailrec
+
+import firrtl._
+import firrtl.ir._
+import firrtl.graph._
+import firrtl.Utils._
+import firrtl.Mappers._
+import firrtl.analyses.InstanceGraph
+
+/** Finds and detects combinational logic loops in a circuit, if any
+  * exist. Returns the input circuit with no modifications.
+  * 
+  * @throws a CombLoopException if a loop is found
+  * @note Input form: Low FIRRTL
+  * @note Output form: Low FIRRTL (identity transform)
+  * @note The pass looks for loops through combinational-read memories
+  * @note The pass cannot find loops that pass through ExtModules
+  * @note The pass will throw exceptions on "false paths"
+  */
+
+object CheckCombLoops extends Pass {
+  def name = "Check Loops"
+
+  class CombLoopException(info: Info, mname: String, cycle: Seq[String]) extends PassException(
+    s"$info: [module $mname] Combinational loop detected:\n" + cycle.mkString("\n"))
+
+  private case class LogicNode(name: String, inst: Option[String] = None, memport: Option[String] = None)
```

Comment:
Probably worth clarifying in a comment what this is used for, and what the fields are for - I know from our previous conversation, but it isn't obvious from just looking at the class.